pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,401 @@
+			}
+			cancel()
+		} else {
+			i.log.Debug("Agent is starting up, not waiting for ipcache while importing policy")
+		}
+	}
+
+	return toPrune
+}
+
+// prunePrefixes removes the CIDR labels from the given set of (resource, prefix) pairs.
+func (i *policyImporter) prunePrefixes(prunePrefixes map[ipcachetypes.ResourceID][]netip.Prefix) {
+	if i.ipc == nil {
+		return
+	}
+	var ipcUpdates []ipcache.MU
+	for resource, oldPrefixes := range prunePrefixes {
+		// Prune all stale prefixes
+		for _, oldPrefix := range oldPrefixes {
+			ipcUpdates = append(ipcUpdates, ipcache.MU{
+				Prefix:   oldPrefix,
+				Resource: resource,
+				Metadata: []ipcache.IPMetadata{labels.Labels{}},
+			})
+		}
+	}
+	if len(ipcUpdates) > 0 {
+		i.log.Info("pruning stale policy CIDR prefix ipcache metadata entries", logfields.Count, len(ipcUpdates))
+		// No need to wait for completion.
+		i.ipc.RemoveMetadataBatch(ipcUpdates...)
+	}
+}
+
+// processUpdates takes a set of one or more policy updates and applies them to
+// the repository. It then regenerates or skips revisions of endpoints as necessary.
+//
+// It also handles prefix allocation in the ipcache when the supplied rules rely on
+// CIDR identities.
+func (i *policyImporter) processUpdates(ctx context.Context, updates []*policytypes.PolicyUpdate) {
+	if len(updates) == 0 {
+		return
+	}
+
+	i.log.Info("Processing policy updates", logfields.Count, len(updates))
+
+	// First, allocate local identities for all prefixes referenced by policies.
+	//
+	// This must happen before the policies are applied to the endpoints. Doing
+	// so prevents traffic from being dropped. Consider a policy that allows
+	// access to 1.1.1.1/32. It only allows traffic to identities that match that selector.
+	// If we were to apply the policy to the endpoint before allocating an identity for
+	// that prefix, traffic to 1.1.1.1/32 may have the world identity, which would be dropped.
+	//
+	// So, we must always perform identity allocation first, then update policy.
+	// The ony exception is if we are starting up, in which case we may proceed.
+	oldPrefixes := i.updatePrefixes(ctx, updates)
+
+	// Apply changes to the repository.
+	//
+	// As we commit to the policy repository, we must also determine the set of identities
+	// to regenerate here. Identities selected by either outgoing or incoming rules
+	// will have to be regenerated.
+	idsToRegen := &set.Set[identity.NumericIdentity]{}
+	startRevision := i.repo.GetRevision()
+	endRevision := startRevision
+	for _, upd := range updates {
+		var regen *set.Set[identity.NumericIdentity]
+
+		// The standard case: we have an owning resource, either a k8s object
+		// or a file on disk.
+		if upd.Resource != "" {
+			regen, endRevision = i.repo.ReplaceByResource(upd.Rules, upd.Resource)
+		} else {
+			// otherwise, this is a local API call, and we are replacing by labels.
+			var removeLabels []labels.LabelArray
```

Comment:
how about `replaceLabels`?

Reply:
The point is that ReplaceByLabels deletes all rules matching `removeLabels`, then inserts the rules. It's on the caller to send a reasonably coherent set of labels, but there's nothing that actually requires them to be "owning".

Reply:
> The point is that ReplaceByLabels deletes all rules matching removeLabels, then inserts the rules.

Ah, this makes sense. Yeah I guess I was a bit confused since some of the labels are presumably going to be re-inserted. In any case, it seems you renamed it to `replaceLabels`. Marking as resolved, I don't feel strongly about this